modules/phase_field/src/materials/ComputeGBMisorientationType.C

```diff
@@ -0,0 +1,315 @@
+}
+
+// Function to convert symmetry matrix to quaternion form
+void
+ComputeGBMisorientationType::rotationSymmetryToQuaternion(const double O[3][3], double q[4])
+{
+  double q4 = 0;
+  if ((1 + O[0][0] + O[1][1] + O[2][2]) > 0)
+  {
+    q4 = sqrt(1 + O[0][0] + O[1][1] + O[2][2]) / 2;
+    q[0] = q4;
+    q[1] = (O[2][1] - O[1][2]) / (4 * q4);
+    q[2] = (O[0][2] - O[2][0]) / (4 * q4);
+    q[3] = (O[1][0] - O[0][1]) / (4 * q4);
+  }
+  else if ((1 + O[0][0] - O[1][1] - O[2][2]) > 0)
+  {
+    q4 = sqrt(1 + O[0][0] - O[1][1] - O[2][2]) / 2;
+    q[0] = (O[2][1] - O[1][2]) / (4 * q4);
+    q[1] = q4;
+    q[2] = (O[1][0] + O[0][1]) / (4 * q4);
+    q[3] = (O[0][2] + O[2][0]) / (4 * q4);
+  }
+  else if ((1 - O[0][0] + O[1][1] - O[2][2]) > 0)
+  {
+    q4 = sqrt(1 - O[0][0] + O[1][1] - O[2][2]) / 2;
+    q[0] = (O[0][2] - O[2][0]) / (4 * q4);
+    q[1] = (O[1][0] + O[0][1]) / (4 * q4);
+    q[2] = q4;
+    q[3] = (O[2][1] + O[1][2]) / (4 * q4);
+  }
+  else if ((1 - O[0][0] - O[1][1] + O[2][2]) > 0)
+  {
+    q4 = sqrt(1 - O[0][0] - O[1][1] + O[2][2]) / 2;
+    q[0] = (O[1][0] - O[0][1]) / (4 * q4);
+    q[1] = (O[0][2] + O[2][0]) / (4 * q4);
+    q[2] = (O[2][1] + O[1][2]) / (4 * q4);
+    q[3] = q4;
+  }
+}
+
+// Function to define the symmetry operator
+void
+ComputeGBMisorientationType::defineSymmetryOperator(double *** sym)
+{
+  // grow by number of symmetric operators
+  (*sym) = new double *[o_sym];
+
+  // grow for symmetry quaternion vectors
+  for (int o = 0; o < o_sym; o++)
+    (*sym)[o] = new double[4];
+
+  // buffer for quaternion
+  double q[4];
+
+  // cubic symmetry
+  double sym_rotation[24][3][3] = {
+      {{1, 0, 0}, {0, 1, 0}, {0, 0, 1}},    {{1, 0, 0}, {0, -1, 0}, {0, 0, -1}},
+      {{1, 0, 0}, {0, 0, -1}, {0, 1, 0}},   {{1, 0, 0}, {0, 0, 1}, {0, -1, 0}},
+      {{-1, 0, 0}, {0, 1, 0}, {0, 0, -1}},  {{-1, 0, 0}, {0, -1, 0}, {0, 0, 1}},
+      {{-1, 0, 0}, {0, 0, -1}, {0, -1, 0}}, {{-1, 0, 0}, {0, 0, 1}, {0, 1, 0}},
+      {{0, 1, 0}, {-1, 0, 0}, {0, 0, 1}},   {{0, 1, 0}, {0, 0, -1}, {-1, 0, 0}},
+      {{0, 1, 0}, {1, 0, 0}, {0, 0, -1}},   {{0, 1, 0}, {0, 0, 1}, {1, 0, 0}},
+      {{0, -1, 0}, {1, 0, 0}, {0, 0, 1}},   {{0, -1, 0}, {0, 0, -1}, {1, 0, 0}},
+      {{0, -1, 0}, {-1, 0, 0}, {0, 0, -1}}, {{0, -1, 0}, {0, 0, 1}, {-1, 0, 0}},
+      {{0, 0, 1}, {0, 1, 0}, {-1, 0, 0}},   {{0, 0, 1}, {1, 0, 0}, {0, 1, 0}},
+      {{0, 0, 1}, {0, -1, 0}, {1, 0, 0}},   {{0, 0, 1}, {-1, 0, 0}, {0, -1, 0}},
+      {{0, 0, -1}, {0, 1, 0}, {1, 0, 0}},   {{0, 0, -1}, {-1, 0, 0}, {0, 1, 0}},
+      {{0, 0, -1}, {0, -1, 0}, {-1, 0, 0}}, {{0, 0, -1}, {1, 0, 0}, {0, -1, 0}}};
+
+  // initialize global operator
+  for (int o = 0; o < o_sym; o++)
+  {
+    rotationSymmetryToQuaternion(sym_rotation[o], q);
+    for (int i = 0; i < 4; i++)
+      (*sym)[o][i] = q[i];
+  }
+}
+
+// Function to convert grain Euler angles to quaternion vector
+void
+ComputeGBMisorientationType::eulerOrientationToQuaternion(int grain_id)
```

Comment:
We already have code to do this in the `EulerAngles`.. which you should be using to hold Euler angles.